


<div class="jp-Notebook" data-jp-theme-light="true" data-jp-theme-name="JupyterLab Light">
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<h2 id="Introduction">Introduction<a class="anchor-link" href="#Introduction">¶</a></h2><p>In this introductory notebook for VoyagerPy, we demonstrate basic exploratory data analysis (<em>EDA</em>) of spatial transcriptomics data. Basic knowledge of Python is assumed.
We furthermore assume that <a href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a> is installed.</p>
<p>This notebook showcases the packages with a Visium spatial gene expression system dataset, downloaded from the 10X website, in the <a href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/output/overview">Space Ranger output format</a>. The technology was chosen due to its popularity, and therefore the availability of numerous publicly available datasets for analysis (Moses and Pachter 2022).</p>
<p>VoyagerPy was developed with the goal of facilitating the use of geospatial methods in spatial genomics. However, this notebook is introductory to the package and is restricted to non-spatial scRNA-seq EDA with the Visium dataset.</p>
<p><strong>Note</strong>: This notebook has an <a href="https://pachterlab.github.io/voyager/articles/visium_10x.html">accompanying vignette</a> in R, so we try to match the results as well as we can.</p>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>We start by loading the basic packages</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [1]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">geopandas</span> <span class="k">as</span> <span class="nn">gpd</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">voyagerpy</span> <span class="k">as</span> <span class="nn">vp</span>

<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>

<span class="c1"># We set the dpi to get clearer figures</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">'figure.dpi'</span><span class="p">]</span> <span class="o">=</span> <span class="mi">120</span>  <span class="c1"># The default is 100</span>

<span class="c1"># Turn on matplotlib interactive mode so we don't need to explicitly call plt.show()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ion</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child jp-OutputArea-executeResult">
<div class="jp-OutputPrompt jp-OutputArea-prompt">Out[1]:</div>
<div class="jp-RenderedText jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/plain">
<pre>&lt;contextlib.ExitStack at 0x7f87d836ef10&gt;</pre>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<h2 id="Downloading-the-data">Downloading the data<a class="anchor-link" href="#Downloading-the-data">¶</a></h2><p>We download the raw count data from the 10X website. These are two gzipped tar archives containing the unfiltered gene count matrix and the spatial information. Thus, we unzip the files.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [2]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">import</span> <span class="nn">pathlib</span>
<span class="kn">import</span> <span class="nn">json</span>

<span class="n">outs_dir</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="s1">'data/visium_10x/outs'</span><span class="p">)</span>
<span class="n">outs_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">root_dir</span> <span class="o">=</span> <span class="p">(</span><span class="n">outs_dir</span> <span class="o">/</span> <span class="s1">'..'</span><span class="p">)</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>

<span class="c1"># Download the gene count matrix</span>
<span class="n">tar_path_ob</span> <span class="o">=</span> <span class="n">root_dir</span> <span class="o">/</span> <span class="s1">'visium_ob.tar.gz'</span>
<span class="n">url_reads</span> <span class="o">=</span> <span class="s2">"https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_raw_feature_bc_matrix.tar.gz"</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">tar_path_ob</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url_reads</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">tar_path_ob</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s1">'wb'</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">content</span><span class="p">)</span>

<span class="c1"># Download the spatial information</span>
<span class="n">tar_path_sp</span> <span class="o">=</span>  <span class="n">root_dir</span> <span class="o">/</span> <span class="s1">'visium_ob_spatial.tar.gz'</span>
<span class="n">url_spatial</span> <span class="o">=</span> <span class="s2">"https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_spatial.tar.gz"</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">tar_path_sp</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url_spatial</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">tar_path_sp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s1">'wb'</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">content</span><span class="p">)</span>

<span class="c1"># Decompress the downloaded files</span>
<span class="o">!</span>tar<span class="w"> </span>-xvf<span class="w"> </span><span class="nv">$tar_path_ob</span><span class="w"> </span>-C<span class="w"> </span><span class="nv">$outs_dir</span><span class="w"> </span>
<span class="o">!</span>tar<span class="w"> </span>-xvf<span class="w"> </span><span class="nv">$tar_path_sp</span><span class="w"> </span>-C<span class="w"> </span><span class="nv">$outs_dir</span><span class="w"> </span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain">
<pre>x raw_feature_bc_matrix/
x raw_feature_bc_matrix/features.tsv.gz
x raw_feature_bc_matrix/barcodes.tsv.gz
x raw_feature_bc_matrix/matrix.mtx.gz
x spatial/
x spatial/spatial_enrichment.csv
x spatial/tissue_lowres_image.png
x spatial/tissue_positions.csv
x spatial/detected_tissue_image.jpg
x spatial/aligned_fiducials.jpg
x spatial/scalefactors_json.json
x spatial/tissue_hires_image.png
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>This is what the layout of the <code>outs</code> directory looks like. The outputs in the spatial directory is explained <a href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/output/spatial">here on the 10X website</a>.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [3]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">print_dir_content</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">d</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">' '</span><span class="o">*</span><span class="p">(</span><span class="n">d</span><span class="o">*</span><span class="mi">2</span><span class="p">)</span> <span class="o">+</span> <span class="n">p</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">'/</span><span class="se">\n</span><span class="s1">'</span> <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">is_dir</span><span class="p">()</span> <span class="k">else</span> <span class="s1">'</span><span class="se">\n</span><span class="s1">'</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">is_dir</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">sub</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">iterdir</span><span class="p">()):</span>
            <span class="n">print_dir_content</span><span class="p">(</span><span class="n">sub</span><span class="p">,</span> <span class="n">d</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>

<span class="n">has_tree</span> <span class="o">=</span> <span class="o">!</span><span class="nb">command</span><span class="w"> </span>-v<span class="w"> </span>tree
<span class="k">if</span> <span class="n">has_tree</span><span class="p">:</span>
    <span class="o">!</span>tree<span class="w"> </span><span class="nv">$outs_dir</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">print_dir_content</span><span class="p">(</span><span class="n">outs_dir</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain">
<pre><span class="ansi-blue-intense-fg ansi-bold">data/visium_10x/outs</span>
├── <span class="ansi-blue-intense-fg ansi-bold">raw_feature_bc_matrix</span>
│   ├── <span class="ansi-red-intense-fg ansi-bold">barcodes.tsv.gz</span>
│   ├── <span class="ansi-red-intense-fg ansi-bold">features.tsv.gz</span>
│   └── <span class="ansi-red-intense-fg ansi-bold">matrix.mtx.gz</span>
└── <span class="ansi-blue-intense-fg ansi-bold">spatial</span>
    ├── <span class="ansi-magenta-intense-fg ansi-bold">aligned_fiducials.jpg</span>
    ├── <span class="ansi-magenta-intense-fg ansi-bold">detected_tissue_image.jpg</span>
    ├── scalefactors_json.json
    ├── spatial_enrichment.csv
    ├── tissue_hires_image.png
    ├── tissue_lowres_image.png
    ├── tissue_positions.csv
    └── tissue_positions_list.csv

3 directories, 11 files
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>The <code>tissue_hires_image.png</code> file is a relatively high resolution image of the tissue, but not full resolution. The <code>tissue_lowres_image.png</code> file is a low resolution image of the tissue, suitable for quick plotting.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [4]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">im</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="n">outs_dir</span><span class="o">/</span><span class="s1">'spatial'</span> <span class="o">/</span> <span class="s1">'tissue_hires_image.png'</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">im</span><span class="p">,</span> <span class="n">origin</span><span class="o">=</span><span class="s1">'lower'</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_8_0.png"/>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>The array of dots surrounding the tissue is the fiducials. These are used for aligning the image to the positions of the Visium spots, so gene expression can be matched to spatial locations. The alignment of the fiducials are shown in <code>aligned_fiducials.jpg</code>. Space Ranger can automatically detect which spots are in tissue. These spots are highlighted in <code>detected_tissue_image.jpg</code> and have <code>in_tissue == 1</code> in <code>tissue_positions.csv</code>.</p>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>The <code>scale_factors_json.json</code> describes how we go from full-res coordinates to the lower resolution coordinates. It also contains the sizes of the spots and fiducials.</p>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p><code>spot_diameter_fullres</code> is the diameter of each Visium spot in the full resolution pixel space. The scalars <code>tissue_hires_scalef</code> and <code>tisse_lowres_scalef</code> are the ratios of the <em>hires</em> and <em>lowres</em> images to the <em>fullres</em> image. <code>fiducial_diameter_fullres</code> is the diameter of each fiducial in full resolution.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [5]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">((</span><span class="n">outs_dir</span> <span class="o">/</span> <span class="s1">'spatial'</span> <span class="o">/</span> <span class="s1">'scalefactors_json.json'</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child jp-OutputArea-executeResult">
<div class="jp-OutputPrompt jp-OutputArea-prompt">Out[5]:</div>
<div class="jp-RenderedText jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/plain">
<pre>{'tissue_hires_scalef': 0.2,
 'tissue_lowres_scalef': 0.06,
 'fiducial_diameter_fullres': 118.91545,
 'spot_diameter_fullres': 73.61433}</pre>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>The file <code>tissue_positions.csv</code> contains information for the Visium spots in the image;</p>
<ul>
<li>barcode</li>
<li><code>in_tissue</code>: whether the barcode is covered by tissue (1) or not (0), as detected by Space Ranger, or manually annotated in the Loupe browser.</li>
<li><code>array_row</code> / <code>array_col</code>: The grid position of the spot on the Visium slide.</li>
<li><code>pxl_row_in_fullres</code> / <code>pxl_col_in_fullres</code>: The pixel position of the spots in the full resolution image.</li>
</ul>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [6]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">outs_dir</span> <span class="o">/</span> <span class="s1">'spatial'</span> <span class="o">/</span> <span class="s1">'tissue_positions.csv'</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child jp-OutputArea-executeResult">
<div class="jp-OutputPrompt jp-OutputArea-prompt">Out[6]:</div>
<div class="jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>barcode</th>
<th>in_tissue</th>
<th>array_row</th>
<th>array_col</th>
<th>pxl_row_in_fullres</th>
<th>pxl_col_in_fullres</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>ACGCCTGACACGCGCT-1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>8668</td>
<td>1102</td>
</tr>
<tr>
<th>1</th>
<td>TACCGATCCAACACTT-1</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>8611</td>
<td>1200</td>
</tr>
<tr>
<th>2</th>
<td>ATTAAAGCGGACGAGC-1</td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>8554</td>
<td>1102</td>
</tr>
<tr>
<th>3</th>
<td>GATAAGGGACGATTAG-1</td>
<td>0</td>
<td>1</td>
<td>3</td>
<td>8498</td>
<td>1200</td>
</tr>
<tr>
<th>4</th>
<td>GTGCAAATCACCAATA-1</td>
<td>0</td>
<td>0</td>
<td>4</td>
<td>8441</td>
<td>1102</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>The file <code>spatial_enrichment.csv</code> contains some information for the genes, e.g. Moran's I and its p-value.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [7]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">outs_dir</span> <span class="o">/</span> <span class="s1">'spatial'</span> <span class="o">/</span> <span class="s1">'spatial_enrichment.csv'</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child jp-OutputArea-executeResult">
<div class="jp-OutputPrompt jp-OutputArea-prompt">Out[7]:</div>
<div class="jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>Feature ID</th>
<th>Feature Name</th>
<th>Feature Type</th>
<th>I</th>
<th>P value</th>
<th>Adjusted p value</th>
<th>Feature Counts in Spots Under Tissue</th>
<th>Median Normalized Average Counts</th>
<th>Barcodes Detected per Feature</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>ENSMUSG00000001023</td>
<td>S100a5</td>
<td>Gene Expression</td>
<td>0.770905</td>
<td>0.0</td>
<td>0.0</td>
<td>9019</td>
<td>15.848669</td>
<td>1021</td>
</tr>
<tr>
<th>1</th>
<td>ENSMUSG00000019874</td>
<td>Fabp7</td>
<td>Gene Expression</td>
<td>0.698735</td>
<td>0.0</td>
<td>0.0</td>
<td>13462</td>
<td>20.679932</td>
<td>1170</td>
</tr>
<tr>
<th>2</th>
<td>ENSMUSG00000002985</td>
<td>Apoe</td>
<td>Gene Expression</td>
<td>0.694521</td>
<td>0.0</td>
<td>0.0</td>
<td>67509</td>
<td>76.635169</td>
<td>1184</td>
</tr>
<tr>
<th>3</th>
<td>ENSMUSG00000025739</td>
<td>Gng13</td>
<td>Gene Expression</td>
<td>0.658575</td>
<td>0.0</td>
<td>0.0</td>
<td>5260</td>
<td>8.803694</td>
<td>1050</td>
</tr>
<tr>
<th>4</th>
<td>ENSMUSG00000090223</td>
<td>Pcp4</td>
<td>Gene Expression</td>
<td>0.631703</td>
<td>0.0</td>
<td>0.0</td>
<td>45118</td>
<td>25.811125</td>
<td>1133</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>Here we read the Space Ranger output as an AnnData object. Since we have the raw counts, we set <code>raw = True</code>. The count matrix is in <code>.mtx</code> format, and we want to load the <code>lowres</code> image.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell jp-mod-noOutputs">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [8]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">read_10x_visium</span><span class="p">(</span>
    <span class="n">outs_dir</span><span class="p">,</span>
    <span class="n">datatype</span> <span class="o">=</span> <span class="s1">'mtx'</span><span class="p">,</span>
    <span class="n">raw</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">prefix</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">symbol_as_index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span>
    <span class="n">res</span><span class="o">=</span><span class="s1">'lowres'</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>We can use VoyagerPy to display the image stored in the adata object.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [9]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_20_0.png"/>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>The images and coordinates of the Visium spots are aligned. However, according to <a href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/algorithms/imaging">the 10X website</a>, a properly aligned slide image should have the hourglass in the top-left corner and the triangle on the bottom left. The rotation of the image does not affect any of the computations, but VoyagerPy offers a way to rotate/mirror the image and the spot coordinates. Here, we rotate and mirror the image to match the fiducials.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [10]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>

<span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s1">'Original'</span><span class="p">)</span>

<span class="n">vp</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">rotate_img90</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">apply</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">tmp</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s1">'Rotated 90° clockwise - not applied'</span><span class="p">)</span>

<span class="n">vp</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">mirror_img</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">apply</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">tmp</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s1">'Mirror rows - not applied'</span><span class="p">)</span>

<span class="n">vp</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">apply_transforms</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s1">'Transformed image - changes applied'</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_22_0.png"/>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<h2 id="Quality-Control-(QC)">Quality Control (QC)<a class="anchor-link" href="#Quality-Control-(QC)">¶</a></h2>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>We start off by computing some basic QC metrics for the dataset.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell jp-mod-noOutputs">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [11]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">is_mt</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">'symbol'</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">'^mt-'</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
<span class="n">vp</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">add_per_cell_qcmetrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">subsets</span><span class="o">=</span><span class="p">{</span><span class="s1">'mito'</span><span class="p">:</span> <span class="n">is_mt</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>Now, we need a way to show display data with respect to their histological location.
Since we are dealing with a 10X Visium dataset, we know how to access this information. Thus, we add the visium spots as the representative geometry of the barcodes.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell jp-mod-noOutputs">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [12]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">scale</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">get_scale</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">spot_diam</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">"spatial"</span><span class="p">][</span><span class="s2">"scale"</span><span class="p">][</span><span class="s2">"spot_diameter_fullres"</span><span class="p">]</span>

<span class="c1"># Create the polygons to represent the Visium spots</span>
<span class="n">visium_spots</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">to_points</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">"pxl_col_in_fullres"</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">"pxl_row_in_fullres"</span><span class="p">,</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span>
    <span class="n">scale</span><span class="o">=</span><span class="n">scale</span><span class="p">,</span>
    <span class="n">radius</span><span class="o">=</span><span class="n">scale</span> <span class="o">*</span> <span class="n">spot_diam</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># or, equivalently </span>
<span class="n">spots_x</span><span class="p">,</span> <span class="n">spots_y</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">get_spot_coords</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">tissue</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">spots</span> <span class="o">=</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoSeries</span><span class="o">.</span><span class="n">from_xy</span><span class="p">(</span><span class="n">spots_x</span><span class="p">,</span> <span class="n">spots_y</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span><span class="o">.</span><span class="n">buffer</span><span class="p">(</span><span class="n">scale</span> <span class="o">*</span> <span class="n">spot_diam</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">spots</span> <span class="o">==</span> <span class="n">visium_spots</span><span class="p">)</span>
<span class="k">del</span> <span class="n">spots</span>

<span class="c1"># Set the geometry to the visium spots and assign the name "spot_poly"</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">set_geometry</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">geom</span><span class="o">=</span><span class="s2">"spot_poly"</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="n">visium_spots</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>Since we have defined the barcode geometry, we can plot the QC metrics in tissue space.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [13]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">qc_features</span> <span class="o">=</span> <span class="p">[</span><span class="s2">"sum"</span><span class="p">,</span> <span class="s2">"detected"</span><span class="p">,</span> <span class="s2">"subsets_mito_percent"</span><span class="p">]</span>
<span class="n">axs</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">plot_spatial_feature</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">qc_features</span><span class="p">,</span> 
    <span class="n">image_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(),</span>
    <span class="n">subplot_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">8</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s1">'tight'</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_29_0.png"/>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>The percentage of mitochondrial counts (<code>subsets_mito_percent</code>) in spots outside tissue is higher near the tissue, especially on the left. See the figure above.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [14]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="c1"># set the `in_tissue` as a categorical variable.</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'in_tissue'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'in_tissue'</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>

<span class="n">axs</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">plot_barcode_data</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">y</span><span class="o">=</span><span class="n">qc_features</span><span class="p">,</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">'in_tissue'</span><span class="p">,</span>
    <span class="n">ncol</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
    <span class="n">cmap</span><span class="o">=</span><span class="s1">'tab10'</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_31_0.png"/>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>Here we can see see three peaks, apparently histologicaly relevant, but no obvious outliers.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [15]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">plot_barcode_data</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">x</span><span class="o">=</span><span class="s1">'sum'</span><span class="p">,</span> 
    <span class="n">y</span><span class="o">=</span><span class="s1">'subsets_mito_percent'</span><span class="p">,</span> 
    <span class="n">color_by</span><span class="o">=</span><span class="s1">'in_tissue'</span><span class="p">,</span> 
    <span class="n">cmap</span><span class="o">=</span><span class="s1">'tab10'</span><span class="p">,</span>
    <span class="n">contour_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">colors</span><span class="o">=</span><span class="s1">'blue'</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="mi">9</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_33_0.png"/>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>This is unlike scRNA-data. Spots not in tissue have a wide range of mitochondrial percentage. Spots in tissue fall into three clusters in the above plot, seemingly related to histological regions.</p>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>Now, we select only visium spots covered by the tissue. Due to the internals of AnnData, we must copy the sliced object and set the geometry of <code>adata_tissue</code> again.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [16]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">adata_tissue</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">"in_tissue"</span><span class="p">]</span><span class="o">==</span><span class="kc">True</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">vp</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">set_geometry</span><span class="p">(</span><span class="n">adata_tissue</span><span class="p">,</span> <span class="s2">"spot_poly"</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child jp-OutputArea-executeResult">
<div class="jp-OutputPrompt jp-OutputArea-prompt">Out[16]:</div>
<div class="jp-RenderedText jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/plain">
<pre>AnnData object with n_obs × n_vars = 1185 × 32285
    obs: 'in_tissue', 'array_row', 'array_col', 'pxl_row_in_fullres', 'pxl_col_in_fullres', 'sum', 'detected', 'subsets_mito_sum', 'subsets_mito_detected', 'subsets_mito_percent'
    var: 'symbol', 'feature_types'
    uns: 'config', 'spatial', 'metadata'
    obsm: 'geometry'</pre>
</div>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [17]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">ax</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plotting</span><span class="o">.</span><span class="n">plot_barcodes_bin2d</span><span class="p">(</span>
    <span class="n">adata_tissue</span><span class="p">,</span> 
    <span class="n">x</span><span class="o">=</span><span class="s1">'sum'</span><span class="p">,</span> 
    <span class="n">y</span><span class="o">=</span><span class="s1">'detected'</span><span class="p">,</span>
    <span class="n">bins</span><span class="o">=</span><span class="mi">76</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">7</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_37_0.png"/>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>In order to preserve the counts and the log-normalized counts, we save them as layers. This is because we will normalize <code>adata_tissue.X</code> before we perform PCA in this notebook.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell jp-mod-noOutputs">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [18]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="c1"># The original count data</span>
<span class="n">adata_tissue</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">'counts'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_tissue</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="c1"># Log-normalize the adata.X matrix</span>
<span class="n">vp</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">log_norm_counts</span><span class="p">(</span><span class="n">adata_tissue</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">adata_tissue</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">'logcounts'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_tissue</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>Next, we select the top 2000 highly variable genes. We use the <code>model_gene_var</code> to model the variance of the gene expression, decomposing the perceived variance into biological variance and technological variance. The biological variance is what we are interested in.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [19]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">gene_var</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">model_gene_var</span><span class="p">(</span><span class="n">adata_tissue</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">'logcounts'</span><span class="p">],</span> <span class="n">gene_names</span><span class="o">=</span><span class="n">adata_tissue</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
<span class="n">hvgs</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">get_top_hvgs</span><span class="p">(</span><span class="n">gene_var</span><span class="p">)</span>

<span class="c1"># Set the 'highly_variable' column for the genes</span>
<span class="n">adata_tissue</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">'highly_variable'</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">adata_tissue</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">hvgs</span><span class="p">,</span> <span class="s1">'highly_variable'</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedText jp-OutputArea-output" data-mime-type="application/vnd.jupyter.stderr">
<pre>/Users/petur/opt/anaconda3/envs/voyager39/lib/python3.9/site-packages/voyagerpy/utils/hvg.py:66: RuntimeWarning: overflow encountered in power
  possible_vals = np.apply_along_axis(lambda x: sum((_vars - (1.13 * x[0] * _means) / ((_means ** x[1]) + x[0] + 0.001)) ** 2), 1, hits)
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<h2 id="Dimension-reduction-and-clustering">Dimension reduction and clustering<a class="anchor-link" href="#Dimension-reduction-and-clustering">¶</a></h2>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>In the <a href="https://pachterlab.github.io/voyager/articles/visium_10x.html#dimension-reduction-and-clustering">companion vignette</a>, the data is scaled prior to computing the PCA. Thus, we follow suit.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell jp-mod-noOutputs">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [20]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="c1"># scale first, then perform pca</span>

<span class="n">adata_tissue</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata_tissue</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_tissue</span><span class="p">,</span> <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1337</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [21]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">elbow_plot</span><span class="p">(</span><span class="n">adata_tissue</span><span class="p">,</span> <span class="n">ndims</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_45_0.png"/>
</div>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [22]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">ax</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">plot_dim_loadings</span><span class="p">(</span>
    <span class="n">adata_tissue</span><span class="p">,</span> 
    <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> 
    <span class="n">show_symbol</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
    <span class="n">ncol</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> 
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_46_0.png"/>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>Cluster the barcodes in PCA space. We adjusted the <code>n_neighbors</code> parameter to find similar clustering as in the R vignette.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell jp-mod-noOutputs">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [23]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">leidenalg</span> <span class="kn">import</span> <span class="n">ModularityVertexPartition</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span>
    <span class="n">adata_tissue</span><span class="p">,</span> 
    <span class="n">n_pcs</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> 
    <span class="n">use_rep</span><span class="o">=</span><span class="s1">'X_pca'</span><span class="p">,</span> 
    <span class="n">method</span><span class="o">=</span><span class="s1">'gauss'</span><span class="p">,</span> 
    <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">80</span>
<span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span>
    <span class="n">adata_tissue</span><span class="p">,</span> 
    <span class="n">random_state</span><span class="o">=</span><span class="mi">29</span><span class="p">,</span> 
    <span class="n">resolution</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">key_added</span><span class="o">=</span><span class="s1">'cluster'</span><span class="p">,</span>
    <span class="n">partition_type</span><span class="o">=</span><span class="n">ModularityVertexPartition</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [24]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">ax</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">plot_pca</span><span class="p">(</span>
    <span class="n">adata_tissue</span><span class="p">,</span> 
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="mi">7</span><span class="p">),</span> 
    <span class="n">ndim</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> 
    <span class="n">color_by</span><span class="o">=</span><span class="s1">'cluster'</span><span class="p">,</span> 
    <span class="n">cmap</span><span class="o">=</span><span class="s1">'tab10'</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_49_0.png"/>
</div>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [25]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">ax</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">plot_spatial_feature</span><span class="p">(</span>
    <span class="n">adata_tissue</span><span class="p">,</span> 
    <span class="s1">'cluster'</span><span class="p">,</span> 
    <span class="n">barcode_geom</span><span class="o">=</span><span class="s1">'spot_poly'</span><span class="p">,</span> 
    <span class="n">image_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">crop</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="mi">10</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_50_0.png"/>
</div>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [26]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">pl</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">spatial_reduced_dim</span><span class="p">(</span>
    <span class="n">adata_tissue</span><span class="p">,</span> 
    <span class="s2">"X_pca"</span><span class="p">,</span> 
    <span class="n">ncomponents</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> 
    <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> 
    <span class="n">divergent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="mi">7</span><span class="p">),</span>
    <span class="n">image_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">crop</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_51_0.png"/>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>Significant markers for each cluster are obtained as follows. Note, that since the clustering is not identical between <a href="https://bioconductor.org/packages/release/bioc/html/bluster.html">bluster</a> and Scanpy, the marker genes will vary slightly.</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell jp-mod-noOutputs">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [27]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">markers</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">get_marker_genes</span><span class="p">(</span><span class="n">adata_tissue</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="n">cluster</span><span class="o">=</span><span class="s1">'cluster'</span><span class="p">)</span>
<span class="n">marker_genes</span> <span class="o">=</span> <span class="n">markers</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [28]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">adata_tissue</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">marker_genes</span><span class="p">,</span> <span class="p">[</span><span class="s1">'symbol'</span><span class="p">]]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child jp-OutputArea-executeResult">
<div class="jp-OutputPrompt jp-OutputArea-prompt">Out[28]:</div>
<div class="jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>symbol</th>
</tr>
</thead>
<tbody>
<tr>
<th>ENSMUSG00000045573</th>
<td>Penk</td>
</tr>
<tr>
<th>ENSMUSG00000098178</th>
<td>Gm42418</td>
</tr>
<tr>
<th>ENSMUSG00000034730</th>
<td>Adgrb1</td>
</tr>
<tr>
<th>ENSMUSG00000064358</th>
<td>mt-Co3</td>
</tr>
<tr>
<th>ENSMUSG00000032532</th>
<td>Cck</td>
</tr>
<tr>
<th>ENSMUSG00000029223</th>
<td>Uchl1</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [29]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">plot_expression</span><span class="p">(</span>
    <span class="n">adata_tissue</span><span class="p">,</span> 
    <span class="n">marker_genes</span><span class="p">,</span>
    <span class="n">groupby</span><span class="o">=</span><span class="s1">'cluster'</span><span class="p">,</span> 
    <span class="n">show_symbol</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
    <span class="n">layer</span><span class="o">=</span><span class="s1">'logcounts'</span><span class="p">,</span>
    <span class="n">cmap</span><span class="o">=</span><span class="s1">'tab10'</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span><span class="mi">7</span><span class="p">),</span>
    <span class="n">scatter_points</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_55_0.png"/>
</div>
</div>
</div>
</div>
</div>
<div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea"><div class="jp-InputPrompt jp-InputArea-prompt">
</div><div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">
<p>These genes show some interesting patterns in spatial context:</p>
</div>
</div>
</div>
</div><div class="jp-Cell jp-CodeCell jp-Notebook-cell">
<div class="jp-Cell-inputWrapper">
<div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">
</div>
<div class="jp-InputArea jp-Cell-inputArea">
<div class="jp-InputPrompt jp-InputArea-prompt">In [30]:</div>
<div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">
<div class="CodeMirror cm-s-jupyter">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">vp</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">plot_spatial_feature</span><span class="p">(</span>
    <span class="n">adata_tissue</span><span class="p">,</span>
    <span class="n">marker_genes</span><span class="p">,</span>
    <span class="n">ncol</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">layer</span><span class="o">=</span><span class="s1">'logcounts'</span><span class="p">,</span>
    <span class="n">subplot_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
        <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="mi">7</span><span class="p">),</span>
        <span class="n">layout</span><span class="o">=</span><span class="s1">'constrained'</span><span class="p">,</span>
    <span class="p">),</span>
    <span class="n">image_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">crop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="jp-Cell-outputWrapper">
<div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">
</div>
<div class="jp-OutputArea jp-Cell-outputArea">
<div class="jp-OutputArea-child">
<div class="jp-OutputPrompt jp-OutputArea-prompt"></div>
<div class="jp-RenderedImage jp-OutputArea-output">
<img class="" src="/voyagerpy/img/visium_10x_files/visium_10x_57_0.png"/>
</div>
</div>
</div>
</div>
</div>
</div>

